### PR TITLE
MPP-206  : in page v2

### PR DESCRIPTION
**Pull request by Francois-Gomis**
### Reason for change

Implement in page V2 

[Linear task](https://linear.app/almapay/issue/MPP-206/)

### Code changes

Add javascript for InPage
Change Pay controller for ajax request.
Cancel order and payment if customer leave Inpage

### How to test

Configure in page in back office and create payment with P2X P3X or P4X
Test pay later - credit - pay now for normal flow

### Checklist for authors and reviewers

- [ X] The title of the PR uses business wording, not technical jargon, for the changelog readers to understand it
- [X ] The PR implements the changes asked in the referenced task / issue
- [ X] The automated tests are compliant with the [testing strategy](https://www.notion.so/almapay/Backend-testing-strategy-06c642cec1bf47b9b8feca3a91ea8d4a)
- [ X] The tests are relevant, and cover the corner/error cases, not only the happy path
- [X ] You understand the impact of this PR on existing code/features

### Non applicable
- [ ] The changes include adequate logging and Datadog traces
- [ ] Documentation is updated (API, developer documentation, ADR, Notion...)

### CODE REVIEW

File: Controller/Payment/CancelInPagePayment.php [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        $this->orderHelper = $orderHelper;
+        $this->almaClient = $almaClient;
+    }
+
+    /**
+     * @return Json
+     */
+    public function execute(): Json
+    {
+        $response = $this->jsonFactory->create();
```
Comment (ClaireAlma) [inline Controller/Payment/CancelInPagePayment.php:62]: manque de retours à la ligne

File: Controller/Payment/CancelInPagePayment.php [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        $order = $this->session->getLastRealOrder();
+        if (!$order) {
+            $this->error($response, 'No Order in session');
+            return $response;
+        }
+        if (!$order->canCancel()) {
+            $this->error($response, 'Order can not be cancelled');
+            return $response;
+        }
+        $order->cancel()->addStatusToHistory(Order::STATE_CANCELED, 'In Page Payment canceled by customer');
```
Comment (ClaireAlma) [inline Controller/Payment/CancelInPagePayment.php:72]: typo error on State_Canceled

File: Controller/Payment/Pay.php [hunk 3]
```diff
@@ -67,29 +120,106 @@ public function execute()
...
+                return $this->errorProcess($order, 'Error: posted payment plan key and order payment plan key are not the same', $paymentID);
+            }
+        }
+
+
+        if ($this->paymentPlansHelper->inPageIsAllowed($paymentPlanKey)) {
+            $response = $this->jsonFactory->create();
+            $response->setData(['error' => false, 'paymentId' => $paymentID]);
+            return $response;
+        } else {
```
Comment (ClaireAlma) [inline Controller/Payment/Pay.php:163]: pas besoin du else

File: Controller/Payment/Pay.php [hunk 4]
```diff
@@ -67,29 +120,106 @@ public function execute()
...
+     * Get request content.
+     *
+     * @return string
+     * @throws InPagePayException
+     */
+    private function getRequestPaymentPlanKey(): string
+    {
+        $requestContent = json_decode($this->request->getContent(), true);
+        if (
+            !isset($requestContent) ||
```
Comment (ClaireAlma) [inline Controller/Payment/Pay.php:180]: mettre le || en debut de ligne

File: Helpers/PaymentPlansHelper.php [hunk 5]
```diff
@@ -93,6 +93,19 @@ public function paymentTriggerIsAllowed(): bool
...
 
+    /**
+     * Check if in page is allowed for a fee plan
+     *
+     * @param string $planKey
+     * @return bool
+     */
+    public function inPageIsAllowed(string $planKey): bool
+    {
+        //TODO create test isPayNow isInstallment isDeferred IsCredit in plan helper and refactor
```
Comment (ClaireAlma) [inline Helpers/PaymentPlansHelper.php:104]: tu as laisser un todo

File: Test/Unit/Controller/Payment/CancelInPagePaymentTest.php [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+    public function setUp(): void
+    {
+        $this->logger = $this->createMock(Logger::class);
+        $this->jsonFactory = $this->createMock(JsonFactory::class);
+        $this->checkoutSession = $this->createMock(Session::class);
+        $this->orderHelper = $this->createMock(OrderHelper::class);
+        $this->almaClient = $this->createMock(AlmaClient::class);
+    }
+    private function getConstructorDependency(): array
```
Comment (ClaireAlma) [inline Test/Unit/Controller/Payment/CancelInPagePaymentTest.php:36]: retour a la ligne entre les functions

File: Controller/Payment/CancelInPagePayment.php (hunk 2 above)
Comment (Francois-Gomis) [reply Controller/Payment/CancelInPagePayment.php:72]: it's a default magento State 
    public const STATE_CANCELED = 'canceled';

File: Controller/Payment/Pay.php (hunk 3 above)
Comment (Francois-Gomis) [reply Controller/Payment/Pay.php:163]: merci

File: etc/adminhtml/system.xml [hunk 7]
```diff
@@ -140,7 +140,7 @@
                            showInStore="1">
                         <label>Activate in-page checkout</label>
                         <source_model>Magento\Config\Model\Config\Source\Yesno</source_model>
-                        <comment>Activate this setting if you want a in-page checkout</comment>
+                        <comment>Activate this setting if you want a in-page checkout for Pay Now, P2X, P3X and P4X</comment>
```
Comment (ClaireAlma) [inline etc/adminhtml/system.xml:143]: an*

**Review by joyet-simon — APPROVED** [review]

**Pull request merged**